providers/gandi.go

```diff
@@ -0,0 +1,278 @@
+package providers
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/rancher/external-dns/dns"
+	gandi "github.com/prasmussen/gandi-api/client"
+	gandiDomain "github.com/prasmussen/gandi-api/domain"
+	gandiZone "github.com/prasmussen/gandi-api/domain/zone"
+	gandiZoneVersion "github.com/prasmussen/gandi-api/domain/zone/version"
+	gandiRecord "github.com/prasmussen/gandi-api/domain/zone/record"
+)
+
+type GandiHandler struct {
+	record      *gandiRecord.Record
+	zone        *gandiZone.ZoneInfoBase
+	zoneVersion *gandiZoneVersion.Version
+	root        string
+	zoneDomain  string
+	zoneSuffix  string
+	sub         string
+}
+
+func init() {
+	gandiHandler := &GandiHandler{}
+
+	apiKey := os.Getenv("GANDI_APIKEY")
+	if len(apiKey) == 0 {
+		logrus.Infof("GANDI_APIKEY is not set, skipping init of %s provider", gandiHandler.GetName())
+		return
+	}
+
+	systemType := gandi.Production
+	testing := os.Getenv("GANDI_TESTING")
+	if len(testing) != 0 {
+		logrus.Infof("GANDI_TESTING is set, using testing platform")
+		systemType = gandi.Testing
+	}
+
+	if err := RegisterProvider("gandi", gandiHandler); err != nil {
+		logrus.Fatal("Could not register Gandi provider")
+	}
+
+	client := gandi.New(apiKey, systemType)
+	gandiHandler.record = gandiRecord.New(client)
+	gandiHandler.zoneVersion = gandiZoneVersion.New(client)
+
+	root := strings.TrimSuffix(dns.RootDomainName, ".")
+	split_root := strings.Split(root, ".")
+	split_zoneDomain := split_root[len(split_root)-2:len(split_root)]
+	zoneDomain := strings.Join(split_zoneDomain, ".")
+	gandiHandler.zoneDomain = zoneDomain
+	gandiHandler.zoneSuffix = fmt.Sprintf(".%s", zoneDomain)
+	gandiHandler.sub = strings.TrimSuffix(root, zoneDomain)
+
+	domain := gandiDomain.New(client)
+	domainInfo, err := domain.Info(zoneDomain)
+	if err != nil {
```

Comment:
Same for the blocks 61-63, 68-70 - would make sense to move them up. So if the error happens, we exit earlier before doing an unneeded settings on gandiHandler.